community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointScheduler.java

```diff
@@ -78,12 +103,16 @@ public boolean getAsBoolean()
             return !checkPointing;
         }
     };
+    private volatile Throwable[] failures = new Throwable[MAX_CONSECUTIVE_FAILURES_TOLERANCE];
```

Comment:
Since this field is never re-assigned, I believe it could be made `final` instead of `volatile`, we still have the volatile `consecutiveFailures` field for the barriers we need.

On the topic of the volatile `consecutiveFailures`, if this can be accessed concurrently, we should make sure that we use atomic operators for incrementing it (through `AtomicIntFieldUpdater`).

On the topic of the `failures` field never being re-assigned: who is responsible for clearing the contents of this field when there is a successful checkpoint? I don't think we want to leave old exceptions around in this array.